src/storage/src/table/state_table.rs

```diff
@@ -0,0 +1,133 @@
+use crate::error::StorageResult;
+use crate::monitor::StateStoreMetrics;
+use crate::{Keyspace, StateStore};
+
+/// `CellBasedTable` is the interface accessing relational data in KV(`StateStore`) with encoding
+/// format: [keyspace | pk | `column_id` (4B)] -> value.
+/// if the key of the column id does not exist, it will be Null in the relation
+pub struct StateTable<S: StateStore> {
+    keyspace: Keyspace<S>,
+
+    /// `ColumnDesc` contains strictly more info than `schema`.
+    column_descs: Vec<ColumnDesc>,
+
+    /// Ordering of primary key (for assertion)
+    order_types: Vec<OrderType>,
+
+    /// Serializer to serialize keys from input rows
+    key_serializer: OrderedRowSerializer,
+
+    /// buffer key/values
+    mem_table: MemTable,
+
+    /// Relation layer
+    cell_based_table: CellBasedTable<S>,
+}
+impl<S: StateStore> StateTable<S> {
+    pub fn new(
+        keyspace: Keyspace<S>,
+        column_descs: Vec<ColumnDesc>,
+        order_types: Vec<OrderType>,
+    ) -> Self {
+        let cell_based_keyspace = keyspace.clone();
+        let cell_based_column_descs = column_descs.clone();
+        Self {
+            keyspace,
+            column_descs,
+            order_types: order_types.clone(),
+            key_serializer: OrderedRowSerializer::new(order_types),
+            mem_table: MemTable::new(),
+            cell_based_table: CellBasedTable::new_adhoc(
+                cell_based_keyspace,
+                cell_based_column_descs,
+                Arc::new(StateStoreMetrics::unused()),
+            ),
+        }
+    }
+
+    /// read methods
+    pub async fn get_row(&self, pk: &Row, epoch: u64) -> StorageResult<Option<Row>> {
+        todo!()
+    }
+
+    pub async fn get_row_by_scan(&self, pk: &Row, epoch: u64) -> StorageResult<Option<Row>> {
+        todo!()
+    }
+
+    /// write methods
+    pub async fn insert(&mut self, pk: Row, value: Row) -> StorageResult<()> {
+        todo!()
+    }
+
+    pub async fn delete(&mut self, pk: Row, old_value: Row) -> StorageResult<()> {
+        todo!()
+    }
+
+    pub async fn update(&mut self, pk: Row, old_value: Row, new_value: Row) -> StorageResult<()> {
+        todo!()
+    }
+
+    fn commit(&mut self, new_epoch: u64) -> StorageResult<()> {
+        todo!()
+    }
+
+    pub async fn iter(&self, pk: Row) -> StorageResult<StateTableRowIter<S>> {
+        todo!()
+    }
+}
+pub struct StateTableRowIter<S: StateStore> {
+    keyspace: Keyspace<S>,
+    /// A buffer to store prefetched kv pairs from state store
+    buf: Vec<(Bytes, Bytes)>,
+    /// The idx into `buf` for the next item
+    next_idx: usize,
+    /// A bool to indicate whether there are more data to fetch from state store
+    done: bool,
+    /// An epoch representing the read snapshot
+    epoch: u64,
+    /// Cell-based row deserializer
+    cell_based_row_deserializer: CellBasedRowDeserializer,
+    /// Statistics
+    _stats: Arc<StateStoreMetrics>,
```

Comment:
we should reconsider the fields in `StateTableRowIter`. It should be a merge iterator of `mem_table` nad `cell_based_table`, so I think maybe it only contains the 2 iterators.